New setting to prevent automatically importing dangling indices

pugnascotia opened the pull request:
Part of #48366.

Introduce a new static setting, `gateway.auto_import_dangling_indices`, which prevents dangling indices from being automatically imported.

I've also updated the dangling index docs a little to cover the new setting, and add some scary caveats.

pugnascotia committed:
Add a setting to control dangling index allocation

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into allocate-dangling-indices-flag

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into allocate-dangling-indices-flag

pugnascotia committed:
Fix allocate setting and implement ITs

The new settings didn't work, but now it does, and I've written a couple
of ITs that create dangling indices and check what affect the setting
has.

I'm still fixing the unit tests though.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into allocate-dangling-indices-flag

pugnascotia committed:
Finish fixing unit tests

pugnascotia committed:
Rename the new setting

pugnascotia committed:
WIP - trying to make new setting static

New setting `gateway.auto_import_dangling_indices` doesn't need to be
dynamic. Unfortunatlely, this has broken one of the ITs.

pugnascotia committed:
Add docs for gateway.auto_import_dangling_indices

pugnascotia committed:
Fix ITs

elasticmachine commented:
Pinging @elastic/es-docs (:Docs)

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Recovery)

pugnascotia commented:
Note: there's more to do to satisfy the issue, e.g. an API to list the danglies, but I plan on doing that separately.

pugnascotia committed:
Add missing license

pugnascotia committed:
Fix test

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into allocate-dangling-indices-flag

pugnascotia commented:
@DaveCTurner @original-brownbear ready for a review, if you have time.

ywelsch reviewed (commented):
Talked to @DaveCTurner who will add some comments as well. No  need to revise until then

DaveCTurner reviewed (changes requested):
Thanks @pugnascotia, I left a few suggestions.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into allocate-dangling-indices-flag

pugnascotia committed:
Address review feedback

pugnascotia commented:
Thanks for the reviews, I've pushed 71bcbde1b6f1e646952e1d017649fe6d2c161a3a to address them, specifically:

   * Reverted the docs changes
   * Default the new setting to false
   * Only subscribe `DanglingIndicesState` to cluster state updates if the new setting is true
   * Revert changes to `UnsafeBootstrapAndDetachCommandIT` as they're no longer needed
   * Strip confusing comments from `DanglingIndicesStateTests` and be specify the new setting's value when calling `createDanglingIndicesState()`.

pugnascotia commented:
@ywelsch @DaveCTurner any chance of a re-review?

ywelsch commented on server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -55,19 +56,35 @@
+        }
+    }
+
+    public void setAllocateDanglingIndicesSetting(boolean allocateDanglingIndices) {
```
this method is superfluous now?

pugnascotia replied:
Oops, yes, good catch.

ywelsch commented on server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -171,10 +188,15 @@ private IndexMetaData stripAliases(IndexMetaData indexMetaData) {
      */
-    private void allocateDanglingIndices() {
+    void allocateDanglingIndices() {
+        if (this.allocateDanglingIndices == false) {
```
this condition is superfluous now?

ywelsch commented on server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -55,19 +56,35 @@
 
     private final Map<Index, IndexMetaData> danglingIndices = ConcurrentCollections.newConcurrentMap();
 
+    private boolean allocateDanglingIndices;
```
this field is superfluous now?

ywelsch commented on server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java:
```diff
@@ -35,16 +36,21 @@
-            .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 0)
-            .put(IndexMetaData.SETTING_VERSION_CREATED, Version.CURRENT)
-            .build();
+        .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1)
```
Let's avoid reformatting code unrelated to the PR.

ywelsch commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -0,0 +1,148 @@
+                Settings.builder()
+                    .put("number_of_shards", shardCount)
+                    .put("number_of_replicas", replicaCount)
+                    .put(Store.INDEX_STORE_STATS_REFRESH_INTERVAL_SETTING.getKey(), 0)
```
why is this necessary?

pugnascotia replied:
Ahem...that's copy-and-paste. Can you tell it's the first time I've written an ES IT? 😅

ywelsch commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -0,0 +1,148 @@
+
+    private static final int MIN_DOC_COUNT = 500;
+    private static final int SHARD_COUNT = 1;
+    private static final int REPLICA_COUNT = 2;
```
why have these constants here? If you're always passing the same shard and replica count to `createAndPopulateIndex`, there is no point in having these. I find there to be a lot of ceremony in this class that is not adding anything.

ywelsch commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -0,0 +1,148 @@
+        ensureGreen();
+
+        logger.info("--> indexing sample data");
+        final int numDocs = between(MIN_DOC_COUNT, 1000);
```
no need to add any docs. Let's keep it simple

ywelsch commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -0,0 +1,148 @@
+        }
+
+        indexRandom(true, docs);
+        flush();
```
why flush?

ywelsch commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -0,0 +1,148 @@
+
+        indexRandom(true, docs);
+        flush();
+        assertThat(client().prepareSearch(name).setSize(0).get().getHits().getTotalHits().value, equalTo((long) numDocs));
```
why check that indexing works ok? Not the point of this test, that is tested elsewhere.

ywelsch commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -0,0 +1,148 @@
+        client().admin().indices().prepareStats(name).execute().actionGet();
+    }
+
+    private void deleteIndex(String indexName) {
```
why define an extra method here? just so that one logging line can be reused? Feels a bit over the top.

ywelsch commented on test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -1467,21 +1469,22 @@ public int size() {
+     * @return the name of the stopped node, if a node was found to stop.
      */
-    public synchronized boolean stopRandomDataNode() throws IOException {
+    public synchronized Optional<String> stopRandomDataNode() throws IOException {
```
where is this change used?

ywelsch commented on test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -1902,6 +1905,12 @@ public String startNode(Settings settings) {
         return startNodes(settings).get(0);
     }
 
+    public void startNode(String nodeName) {
```
where is this used? Revert

ywelsch commented on server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -55,19 +56,35 @@
 
     private static final Logger logger = LogManager.getLogger(DanglingIndicesState.class);
 
+    public static final Setting<Boolean> AUTO_IMPORT_DANGLING_INDICES_SETTING = Setting.boolSetting(
```
Let's deprecate this setting right away, so that we can remove it in 8.0 if we want

ywelsch commented on server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -55,19 +56,35 @@
+        this.allocateDanglingIndices = AUTO_IMPORT_DANGLING_INDICES_SETTING.get(clusterService.getSettings());
+
+        if (this.allocateDanglingIndices) {
+            clusterService.addListener(this);
```
with this change, we are now not scanning for new dangling indices anymore and therefore also not warning about the existence of dangling indices on disk. We should keep this in mind when working out a new story for managing dangling indices via an API. Perhaps we should think about still emitting a warning in the logs. Let's add a TODO for this on the meta issue.

ywelsch reviewed (changes requested):
Thanks for the update. I've left some comments.

pugnascotia committed:
Address review feedback

pugnascotia committed:
Address review comments

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into allocate-dangling-indices-flag

pugnascotia commented:
@ywelsch thank you for the review - seems there was quite a bit of cruft to clear up, sorry about that. Notes on the latest changes:

   * The new setting is now deprecated
   * The state class logs a warning if the setting is false, since dangling indices won't be detected or imported
   * Change the disabled case unit test to verify that no cluster state listener is registered, since the `danglingState.allocateDanglingIndices()` method no longer checks the setting value
   * Tidied up the IT

pugnascotia committed:
Checkstyle

ywelsch commented on server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java:
```diff
@@ -46,6 +51,12 @@
 
+    // The setting AUTO_IMPORT_DANGLING_INDICES_SETTING is deprecated, so we must disable
+    // warning checks or all the tests will fail.
+    protected boolean enableWarningsCheck() {
```
@Override

ywelsch commented on server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java:
```diff
@@ -181,7 +191,66 @@ public void testDanglingIndicesStripAliases() throws Exception {
+            createDanglingIndicesState(env, metaStateService, localAllocateDangledIndices, mockClusterService);
+
+            // Check that no listener was registered, because auto-import is disabled
+            verify(mockClusterService, never()).addListener(any());
```
I'm not a big fan of these whitebox tests, given how fragile they are. There is also no test here that checks that `addListener` is ever called when auto-import is enabled? This means that if the dangling indices implementation changes to use another method on clusterservice to register a callback, this test will be pointless.

ywelsch commented on server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java:
```diff
@@ -181,7 +191,66 @@ public void testDanglingIndicesStripAliases() throws Exception {
+        LocalAllocateDangledIndices indexAllocator,
+        ClusterService clusterService
+    ) {
+        return new DanglingIndicesState(env, metaStateService, indexAllocator, clusterService);
```
what's the point of this method? Let's avoid these unnecessary abstractions.

ywelsch commented on server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java:
```diff
@@ -181,7 +191,66 @@ public void testDanglingIndicesStripAliases() throws Exception {
+            MetaStateService metaStateService = new MetaStateService(env, xContentRegistry());
+            LocalAllocateDangledIndices localAllocateDangledIndices = mock(LocalAllocateDangledIndices.class);
+
+            final ClusterService mockClusterService = buildMockClusterService(false);
```
I would find the test easier to understand if we were to explicitly pass the setting `Settings.builder().put(AUTO_IMPORT_DANGLING_INDICES_SETTING.getKey(), false).build()` here instead of the magic boolean.

ywelsch commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -0,0 +1,114 @@
+
+        // Create an index and distribute it across the 3 nodes
+        createAndPopulateIndex(INDEX_NAME, 1, 2);
+        ensureGreen();
```
`createAndPopulateIndex` already has a call to `ensureGreen`.

ywelsch commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -0,0 +1,114 @@
+        assertFalse("Did not expect dangling index " + INDEX_NAME + " to be recovered", indexExists(INDEX_NAME));
+    }
+
+    private void createAndPopulateIndex(String name, int shardCount, int replicaCount) {
```
I would prefer not to have this method, just do the actual calls 

Note that there is also the `createIndex(String name, Settings indexSettings)` on ESIntegTestCase which allows to write this succinctly.

ywelsch commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    private void createAndPopulateIndex(String name, int shardCount, int replicaCount) {
+        logger.info("--> creating test index: {}", name);
```
the logging here is not necessary. Elasticsearch already logs index creations at info level.

ywelsch commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -0,0 +1,114 @@
+        createAndPopulateIndex(INDEX_NAME, 1, 2);
+
+        // Create another index so that once we drop the first index, we
+        // can still assert that the cluster is green.
```
I don't understand this.`ensureGreen` does not require for an index to be present?

ywelsch commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -0,0 +1,114 @@
+
+            @Override
+            public Settings onNodeStopped(String nodeName) throws Exception {
+                logger.info("--> deleting test index: {}", INDEX_NAME);
```
deletions are already logged by ES at info level, no need to log here.

ywelsch commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -0,0 +1,114 @@
+
+        ensureGreen();
+
+        assertTrue("Expected dangling index " + INDEX_NAME + " to be recovered", indexExists(INDEX_NAME));
```
There is no guarantee that the dangling index has been recovered at this point, as that has an asynchronous notification / import process. This line needs to be put into an `assertBusy`

ywelsch commented on test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -1473,7 +1473,8 @@ public int size() {
     /**
-     * Stops a random data node in the cluster. Returns true if a node was found to stop, false otherwise.
+     * Stops a random data node in the cluster and removes it.
+     * @return the name of the stopped node, if a node was found to stop.
```
revert this.

ywelsch commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -0,0 +1,114 @@
+     * the cluster when the recovery setting is disabled.
+     */
+    public void testDanglingIndicesAreNotRecoveredWhenSettingIsDisabled() throws Exception {
+        logger.info("--> starting cluster");
```
what does this logging message add? It's the first line in the test, which is already logged.

ywelsch commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -0,0 +1,114 @@
+
+        // This is so that when then node comes back up, we have a dangling index that could
+        // be recovered, but shouldn't be.
+        logger.info("--> restarted a random node and deleting the index while it's down");
```
The restart is already logged at info level. Maybe just change the comment to say `// Restart node, deleting the index in its absence, so that there is a dangling index to recover`

ywelsch reviewed (changes requested):
Thanks for the update @pugnascotia. I've left more comments.

pugnascotia committed:
Address review comments

pugnascotia commented:
Thanks again @ywelsch for bearing with me.

   * I've replaced the white box tests with a getter for the new setting. I had avoided that before because the only code calling the getter would be test code, but it'll be less fragile.
   * I've simplified the unit tests and ITs.

ywelsch commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -0,0 +1,95 @@
+        final Settings settings = buildSettings(true);
+        internalCluster().startNodes(3, settings);
+
+        createIndex(INDEX_NAME, Settings.builder().put("number_of_shards", 1).put("number_of_replicas", 2).build());
```
does the number of (primary) shards matter? I understand that you want 2 replicas so that every node has a shard copy.

ywelsch commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -0,0 +1,95 @@
+        // amount of time
+        assertFalse(
+            "Did not expect dangling index " + INDEX_NAME + " to be recovered",
+            waitUntil(() -> indexExists(INDEX_NAME), 5, TimeUnit.SECONDS)
```
Let's reduce this to 1 second

ywelsch commented on server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -67,7 +76,18 @@ public DanglingIndicesState(NodeEnvironment nodeEnv, MetaStateService metaStateS
+        }
+    }
+
+    public boolean isAutoImportDanglingIndicesEnabled() {
```
does not need to be public, only accessed by tests

ywelsch reviewed (approved):
LGTM

pugnascotia committed:
Address review feedback

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into allocate-dangling-indices-flag
